### ISSUE TITLE
Release builds can make the (master channel) Flutter build dashboard tree red

### ISSUE THREAD

**Issue by christopherfujino** [flutter/flutter#149219]
1. the internal google roll team created a new flutter/flutter branch, named [flutter-3.23-candidate.10](https://github.com/flutter/flutter/tree/flutter-3.23-candidate.10)
2. Step 1 also triggered a corresponding [branch](https://flutter.googlesource.com/recipes/+/refs/heads/flutter-3.23-candidate.10) to be created on the recipes repo, which then triggered a recipes bundle job to be kicked off to upload the recipe to CIPD ([build](https://ci.chromium.org/ui/p/flutter/builders/prod/recipes-bundler/b8746659936919381953/infra))
3. Step 1 also triggered set of release builds, in dart-internal, under the Linux flutter_release_builder
4. A Linux flutter_release_builder job from #3 tried to schedule a Linux docs_publish build with the flutter-3.23-candidate.10 bundle of the recipes BEFORE step 2 had finished (race condition), and thus failed
5. The build dashboard detected this release build failure as part of its check for post-submit builds on the master channel, and thus determined that the tree should be red (even though this builder does not appear in flutter/flutter's .ci.yaml, AND that it was for a release branch)

**Comment by christopherfujino** [comment]
@stuartmorgan do you have a hypothesis as to why the release build: https://chrome-swarming.appspot.com/task?id=69d9f4bc065fb510&o=true&w=true would be detected by cocoon as a main branch engine failure? Something something pubsub?

**Comment by christopherfujino** [comment]
> My initial guess would be something something incorrect database query.
> 
> What was the exact symptom? The banner at the top was red and listed `docs_publish` as the failing test?

I'm assuming (I never saw it myself). @ditman or @loic-sharma do you recall?

**Comment by christopherfujino** [comment]
Oh, but I think `Linux flutter_release_builder` was listed as the failing target.

**Comment by loic-sharma** [comment]
Yes that's right, https://flutter-dashboard.appspot.com/#/build had a red banner that said that `Linux flutter_release_builder` had failed (however there was no column for `Linux flutter_release_builder`).

**Comment by stuartmorgan** [comment]
> @stuartmorgan do you have a hypothesis as to why the release build: https://chrome-swarming.appspot.com/task?id=69d9f4bc065fb510&o=true&w=true would be detected by cocoon as a main branch engine failure?

That task is referencing `fdca33c5b115e76073373326cbd15bdff4839650` as the commit sha, which is in the Firestore database as a `master` commit—which it also is.

I think there may be a fundamental problem with the database design here. Commits use the sha as their document ID, and list a single branch they are part of, so I'm not seeing a way to express the idea of a commit being on two branches. And then tasks have a commit, but no branch; the build status query just finds every task referencing each commit that it found from querying the commits for the default branch.

So while I haven't traced this through the system all the way, it appears that because the commit was a master commit, any task triggered later from that branch pointing to a `master` commit will be attributed to `master` by Cocoon.

I think to prevent this we would either need to fundamentally reconsider how commits+tasks are tracked for branches, or we would need to make a new commit as part of a new branch. E.g., instead of marking `fdca33c5b115e76073373326cbd15bdff4839650` as the candidate, we would have to make a child commit of `fdca33c5b115e76073373326cbd15bdff4839650` that's the start of the branch, and have the branch point to that.

> 6\. The PR [Change `android_plugin_new_output_dir_test.dart` test description #149198](https://github.com/flutter/flutter/pull/149198) is manually merged, which immediately leads to the tree becoming green (why?)

This part I can definitely answer: the build status finds every task associated with the latest commit to the default branch, and then checks the latest completed status of each of those tasks. So as soon as the latest commit doesn't have `Linux flutter_release_builder`, it will no longer factor into build status calculation.

**Comment by christopherfujino** [comment]
Ohh, I remember this being a bug, that I'm sure is tracked somewhere, and it was fundamentally hard to fix via the GitHub API, at least the last time I chatted with Keyong about it.
